ilastik/applets/base/appletSerializer.py

```diff
@@ -1197,6 +1197,13 @@ def updateWorkingDirectory(self, newdir, olddir):
         if they store relative paths."""
         pass
 
+    def updateLegacyProjectFile(self, file: h5py.File):
```

Comment:
This method is now always called when importing but not necessarily does change things. I'd propose to either check outside this method whether it should be called, or rename it to `maybeUpdateProjectFile` or something like this. And also it could be private as it's only used from _importProjectFile, which itself is already a private method.

Reply:
[The internet](https://softwareengineering.stackexchange.com/questions/365310/how-to-name-a-method-which-may-or-may-not-perform-an-action-depending-on-a-condi) seems to lean towards "just name the function what it does if it meets its conditions". Or something like `ensureProjectFileIsUpToDate`. `updateLegacyEntries`. I'm ok with any.

I'm not a fan of splitting this into
```
if serializer.isLegacyProject(self.currentProjectfile):
  seralizer.updateProjectFile(self.currentProjectFile)
```
because then serializer subclasses have to implement two functions and their relation to each other isn't necessarily obvious besides containing some duplicated code.

I tried to see if the name should be underscored or not, but it didn't seem to matter around these parts (ProjectManager._importProject and _loadProject are both called in Shell._loadProject 🙃 ). Will underscore.

Reply:
I'm fine keeping it as is although I'd prefer some kind of conditional. The logic for this has to be implemented anyhow, but yes, you're right, it would mean implementing a second function which would be more complicated to spot.

Reply:
This is what an implementation with an exit step based on storage version could look like:

```
    # (self.version = "0.2" now bumped to self.version = "0.3")

    def updateLegacyEntries(self, file: File):
        try:
            info_dir = file[self.topGroupName]["infos"]
            storage_version = file[self.topGroupName]["StorageVersion"][()].decode()
        except KeyError:
            return
        
        storage_version = parse_version(storage_version)  # parse() from packaging.version
        if not self._isOutdatedStorage(storage_version):
            return

        if storage_version == parse_version("0.2"):
            # Version 0.2 may contain UrlDatasetInfo, the HBP-mode predecessor of MultiscaleUrlDatasetInfo

            # ... (Check all entries for UrlDatasetInfo)
                    
        del file[self.topGroupName]["StorageVersion"]
        file[self.topGroupName]["StorageVersion"] = self.version.encode()
                
    def _isOutdatedStorage(self, version: Version):
        return version < parse_version(self.version)
```

It does provide an obvious point for where to put analogous changes for future version bumps, but it's also a bunch of extra basically unneeded code...